app/android/runtime_client/src/org/xwalk/app/runtime/extension/XWalkExtensionClient.java

```diff
@@ -142,9 +189,134 @@ public void onMessage(int extensionInstanceID, String message) {
      * @param message the message from JavaScript code.
      */
     public String onSyncMessage(int extensionInstanceID, String message) {
-        return "";
+        String TAG = "Extension-" + mName;
+        Object result = null;
+        try {
+            JSONObject m = new JSONObject(message);
+            String cmd = m.getString("cmd");
+            String memberName = m.getString("name");
+            try {
+                switch (cmd) {
+                    case "invokeNative":
+                        result = reflection.invokeMethod(extensionInstanceID,
+                                 this, memberName, m.getJSONArray("args"));
+                        break;
+
+                    case "getProperty":
+                        result = reflection.getProperty(this, memberName);
+                        break;
+
+                    case "setProperty":
+                        reflection.setProperty(this, memberName, m.get("value"));
+                        break;
+
+                    default:
+                        Log.w(TAG, "Unsupported cmd: " + cmd);
+                        break;
+                }
+            } catch (Exception e) {
+                if (e instanceof IllegalArgumentException) {
+                    logJs(extensionInstanceID, e.toString(), "warn");
+                } else {
+                    Log.w(TAG, "[Failed to access member] " + e.toString());
+                }
+                e.printStackTrace();
+            }
+        } catch (Exception e) {
+            Log.w(TAG, "[Invalid message] " + e.toString());
+            e.printStackTrace();
+        }
+        return (result != null) ? ReflectionHelper.objToJSON(result): "";
+    }
+
+    public void invokeJsCallback(JSONObject callInfo, String key, Object... args) {
+        /* Following message will be sent to JavaScript side:
+         * {
+         *  cmd:"invokeCallback"
+         *  // need to combine the cid and instanceId in the same feild
+         *  callInfo: an object contains the callback information(cid, vid)
+         *  key: String
+         *  args: args
+         * }
+         */
```

Comment:
I prefer to put the comments as function comments, which is saying moving them right above function declaration line. Same for other similar funtcions.